internal/bft/viewchanger.go

```
@@ -512,20 +535,190 @@ func (v *ViewChanger) processViewDataMsg() {
+			}
+			inFlightMetadata := &protos.ViewMetadata{}
+			if err := proto.Unmarshal(vd.InFlightProposal.Metadata, inFlightMetadata); err != nil {
+				panic(fmt.Sprintf("Node was unable to unmarshal the in flight proposal metadata, error: %v", err))
+			}
+
+			if inFlightMetadata.LatestSequence != expectedSequence {
+				possible.noArgument++
+				continue
+			}
+
+			if isSameProposals(vd.InFlightProposal, possible.proposal) {
+				possible.noArgument++
+				possible.preprepared++
+			}
+
+		}
+	}
+
+	// see if there is an in flight proposal that is agreed on
+	agreed := -1
+	for i, possible := range possibleProposals {
+		if possible.preprepared < f+1 { // condition A2 doesn't hold
+			continue
+		}
+		if possible.noArgument < quorum { // condition A1 doesn't hold
+			continue
+		}
+		agreed = i
+		break
+	}
+
+	// condition A holds
+	if agreed != -1 {
+		return true, false, possibleProposals[agreed].proposal
+	}
+
+	return false, false, nil
+}
+
+func isSameProposals(p1, p2 *protos.Proposal) bool {
+	if p1.VerificationSequence != p2.VerificationSequence {
+		return false
+	}
+	if !isEqualBytes(p1.Header, p2.Header) {
+		return false
+	}
+	if !isEqualBytes(p1.Metadata, p2.Metadata) {
+		return false
+	}
+	if !isEqualBytes(p1.Payload, p2.Payload) {
+		return false
+	}
+	return true
+}
+
+func isEqualBytes(a, b []byte) bool {
+	if (a == nil) != (b == nil) { // If one is nil, the other must also be nil.
+		return false
+	}
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func MaxLastDecisionSequence(messages []*protos.ViewData, quorum int, N uint64, verifier api.Verifier) uint64 {
+	max := uint64(0)
+	for _, vd := range messages {
+		err, seq := ValidateLastDecision(vd, quorum, N, verifier)
+		if err != nil {
+			panic(fmt.Sprintf("Node was unable to validate last decision in viewData message, error: %v", err))
```

I don't think we should panic in exported functions such as these. We can return an error instead.

---

same as before...